agent/consul/xdscapacity/capacity.go

```diff
@@ -0,0 +1,203 @@
+	defer close(a.doneCh)
+
+	ws, numProxies, err := a.countProxies()
+	if err != nil {
+		a.cfg.Logger.Error("failed to count proxy services", "error", err)
+	}
+
+	var numServers uint32
+	for {
+		select {
+		case s := <-a.serverCh:
+			numServers = s
+			a.updateMaxSessions(numServers, numProxies)
+		case <-ws.WatchCh(ctx):
+			var count uint32
+			ws, count, err = a.countProxies()
+			if err == nil {
+				numProxies = count
+				a.updateDrainRateLimit(numProxies)
+				a.updateMaxSessions(numServers, numProxies)
+			} else {
+				a.cfg.Logger.Error("failed to count proxy services", "error", err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// SetServerCount updates the number of healthy servers that is used when
+// determining capacity. It is called by the autopilot delegate.
+func (a *Controller) SetServerCount(count uint32) {
+	select {
+	case a.serverCh <- count:
+	case <-a.doneCh:
+	}
+}
+
+func (c *Controller) updateDrainRateLimit(numProxies uint32) {
+	rateLimit := calcRateLimit(numProxies)
+	if rateLimit == c.prevRateLimit {
+		return
+	}
+
+	c.cfg.Logger.Debug("updating drain rate limit", "rate_limit", rateLimit)
+	c.cfg.SessionLimiter.SetDrainRateLimit(rateLimit)
+	c.prevRateLimit = rateLimit
+}
+
+// We dynamically scale the rate at which excess sessions will be drained
+// according to the number of proxies in the catalog.
+//
+// The numbers here are pretty arbitrary (change them if you find better ones!)
+// but the logic is:
+//
+//	0-512 proxies: drain 1 per second
+//	513-2815 proxies: linearly scaled by 1/s for every additional 256 proxies
+//	2816+ proxies: drain 10 per second
+//
+func calcRateLimit(numProxies uint32) rate.Limit {
+	perSecond := math.Floor((float64(numProxies) - 256) / 256)
+
+	if perSecond < 1 {
+		return 1
+	}
+
+	if perSecond > 10 {
+		return 10
+	}
+
+	return rate.Limit(perSecond)
+}
+
+func (a *Controller) updateMaxSessions(numServers, numProxies uint32) {
+	if numServers == 0 || numProxies == 0 {
+		return
+	}
+
+	maxSessions := uint32(math.Ceil((float64(numProxies) / float64(numServers)) * (1 + errorMargin)))
+	if maxSessions == a.prevMaxSessions {
+		return
+	}
+
+	a.cfg.Logger.Debug(
+		"updating max sessions",
+		"max_sessions", maxSessions,
+		"num_servers", numServers,
+		"num_proxies", numProxies,
+	)
+	metrics.SetGauge([]string{"xds", "server", "idealStreamsMax"}, float32(maxSessions))
+	a.cfg.SessionLimiter.SetMaxSessions(maxSessions)
+	a.prevMaxSessions = maxSessions
+}
+
+func (a *Controller) countProxies() (memdb.WatchSet, uint32, error) {
+	store := a.cfg.GetStore()
+
+	ws := memdb.NewWatchSet()
+	ws.Add(store.AbandonCh())
+
+	var count uint32
+	_, csns, err := store.ServiceDump(
+		ws,
+		"",
+		false,
+		structs.WildcardEnterpriseMetaInPartition(acl.WildcardName),
+		structs.DefaultPeerKeyword,
+	)
+	if err != nil {
+		return ws, 0, err
+	}
+	for _, csn := range csns {
+		if csn.Service.Kind.IsProxy() {
+			count++
+		}
+	}
+	return ws, count, nil
```

Comment:
🔔  (non-blocking suggestion)

This should perform quite a bit better when there are many service instances and takes advantage of the fact that we are already tracking counts of services by their kind.

```suggestion
	var count int
	_, usage, err := store.ServiceUsage()
	if err != nil {
		return ws, 0, err
	}

	for kind, kindCount := range usage.ConnectServiceInstance {
		if structs.ServiceKind(kind).IsProxy() {
			count += kindCount
		}
	}
	return ws, count, nil
```

Reply:
Oh nice! I had no idea about this 🙌🏻